rfcs/react-components/convergence/slot-render-as.md

```diff
@@ -0,0 +1,155 @@
+
+This proposes modifying the slot render function API to pass a render function via the slot's `as` prop instead of the `children` prop. This has several benefits, the primary of which is that it allows the user to write a render function for a slot without overriding the default children of a slot.
+
+This is a complementary proposal to bsunderhus's [RFC #27164: slot children render function support](https://github.com/microsoft/fluentui/pull/27164). It provides an alternative way to avoid having a render function override a slot's default `children` prop. This RFC has no dependencies on it, and could be adopted separately if needed.
+
+## Background
+
+Slots accept render functions that allow them to override the component and/or props used when rendering the slot. For more background, see https://react.fluentui.dev/?path=/docs/concepts-developer-customizing-components-with-slots--page#replacing-the-entire-slot
+
+The render function is passed in as the slot's `children` prop.
+
+## Problem statement
+
+A few problems with render functions as they are today:
+
+1. Using `children` as the render function will override any default children provided by the component's internals.
+2. It is confusing that passing a render function as `children` would override the _whole slot_, and not just its children.
+3. The root slot's render function conflicts the use of children as a function, which certain components use for custom rendering of their children.
+   - For example, Tooltip and Field both accept a function as a child of their root, in order to pass props to the children. However, this breaks the ability to use a slot render function for the root slot of those controls.
+   - See: https://react.fluentui.dev/?path=/docs/preview-components-field--default#complex-content-in-a-field
+
+And a less important issue that's addressed in Appendix A of this proposal:
+
+4. The order of the arguments is `(Component, props)`, but one of the primary use cases is to replace the given slot component with something else, in which case you ignore the first argument. You'd rarely/never(?) want to ignore the second argument `props`.
+
+## Proposal
+
+The core proposal in this RFC is to use the `as` prop as the render function instead of `children`.
+
+```jsx
+<>
+  {/* Existing behavior: unintentionally overwrites default children of the button slot. */}
+  <AccordionHeader button={{ children: (Component, props) => <Component {...props} /> }} />
+
+  {/* New behavior, use 'as' instead, and `props` includes the slot's default `children`. */}
+  <AccordionHeader button={{ as: (Component, props) => <Component {...props} /> }} />
+</>
+```
+
+You can still pass in a string to the `as` prop (e.g. `as="a"` to render an `<a>` tag), and it would work as it does today:
+
+- If the slot's type is an intrinsic element like `'button'`, it replaces that element with the given one.
+- If the slot's type is a component like `Button`, it forwards the `as` prop to the component.
```

Comment:
To clarify, the statement in the RFC is _not_ about passing a component to the as prop: ~~`as={MyComponent}`~~. Instead, it's describing what happens in the case where the slot is a component (as specified in `state.components`):
```ts
components: {
  root: 'div', // intrinsic element slot
  label: Label, // 👈 component slot
}
```

What it's saying is that `as={string}` continues to work as it does today for both slot types (no change in behavior), and then it's just clarifying what the existing behavior is:
* For intrinsic element slots: `as={string}` replaces the intrinsic element used to render the slot
* For component slots: `as={string}` gets forwarded to the component's own `as` prop.

I can try to make that clearer in the RFC.

> Can you please clarify how components will be distinguished from render functions?

Setting `as={FunctionComponent}` is _not_ part of this RFC. It might happen to work to use a function component here because it could be used as render functions. But ClassComponent or ForwardRefComponent _won't_ work here because they are classes and not functions. In the future we _could_ add support for those as well. But that change is outside of the scope of this RFC.

> The problem is that internally we were doing `React.createElement(props.as)` ➡️ we will get an unstable reference for an element type ➡️ remounting a part of tree on every re-render.

This RFC is _not_ recommending to call `React.createElement`; it's recommending invoking the render function just like we do today for render functions passed to the `children` prop. So the performance should be no different from having a render function as children.

Reply:
> Setting as={FunctionComponent} is not part of this RFC. It might happen to work to use a function component here because it could be used as render functions. But ClassComponent or ForwardRefComponent won't work here because they are classes and not functions. In the future we could add support for those as well. But that change is outside of the scope of this RFC.

I got that, my point is that it _becomes_ possible and it's confusing:
- `children` are inner content or a render function (currently)
- `as` is a component type or a render function (with your proposal)

The problem is you said, you can't identify what is a component and what is a render function. For example, in components you can use hooks, in render functions ESLint will notice that:

```ts
function Component() {
  React.useCallback(() => {}, [])
}

function renderFn() {
  // 🚨 Hooks cannot be used outside hooks and components
  React.useCallback(() => {}, [])
}
```

There are use cases in Teams when render functions are conditional:

```ts
<Button icon={{ children: isFoo ? renderFn : undefined }} />
// 💥 "Component" is considered as a function and it contains hooks
<Button icon={{ as: isFoo ? Component : undefined }} />
```

Reply:
Okay, due to the possibility of accidentally using a FunctionComponent for the `as` prop (resulting in 💥), then I'd say we can remove Appendix A from the proposal.